controller/src/main/java/io/pravega/controller/store/stream/ZkOrderedStore.java

```diff
@@ -92,30 +95,54 @@
+                .thenCompose(latestCollectionNumber ->
+                        storeHelper.createPersistentSequentialZNode(getEntitySequentialPath(scope, stream, latestCollectionNumber),
                                 entity.getBytes(Charsets.UTF_8))
                                    .thenCompose(positionPath -> {
                                        int position = getPositionFromPath(positionPath);
                                        if (position > rollOverAfter) {
-                                           // if newly created position exceeds rollover limit, we need to delete that entry 
-                                           // and roll over. 
-                                           // 1. delete newly created path
-                                           return storeHelper.deletePath(positionPath, false)
-                                                             // 2. seal latest collection
-                                                             .thenCompose(v -> storeHelper.createZNodeIfNotExist(
-                                                                     getCollectionSealedPath(scope, stream, latestcollectionNum)))
-                                                             // 3. call addEntity recursively
-                                                             .thenCompose(v -> addEntity(scope, stream, entity))
-                                                             // 4. delete empty sealed collection path
-                                                             .thenCompose(orderedPosition -> 
-                                                                     tryDeleteSealedCollection(scope, stream, latestcollectionNum)
-                                                                     .thenApply(v -> orderedPosition));
+                                           // if newly created position exceeds rollover limit, we need to delete that entry
+                                           // and roll over.
+                                           try {
+                                               // prepare transaction: delete newly created path, create new collection and put the entity in
+                                               int newCollectionNumber = latestCollectionNumber + 1;
+
+                                               String collectionPath = getCollectionPath(scope, stream, newCollectionNumber);
+                                               String entitiesPath = getEntitiesPath(scope, stream, newCollectionNumber);
+                                               String entitySequentialPath = getEntitySequentialPath(scope, stream, newCollectionNumber);
+
+                                               CuratorOp op1 = storeHelper.getClient().transactionOp().delete().forPath(positionPath);
+                                               CuratorOp op2 = storeHelper.getClient().transactionOp().create().forPath(collectionPath);
```

Comment:
Yes you're right. It is due to a limitation of ZooKeeper that we cannot create if not exist. If we want to tolerate this, then we have to do some retry logics as it is done in Curator.

Here is a quick thought about the original purpose and a smaller patch.

https://github.com/pravega/pravega/issues/3570#issuecomment-524561820
https://github.com/pravega/pravega/issues/3570#issuecomment-524562760

Reply:
So the transaction would fail in such a situation if the node exists. Which would continue to fail even if we call it recursively, unless create is replaced with update in subsequent iteration attempt.

Reply:
yeah, we need to make that change otherwise it will not result in correct behaviour.

Reply:
What do you think on the other approach linked above and posted in the corresponding issue? I think the original purpose we introduce multi-op is to improve readability by reduce recursive calls. The other approach fits the purpose in a more neat way. Instead of we JUST introduce transaction but twist quite a bit.

Reply:
In other words, we originally want to improve readability, but if we have to introduce too many changes because of we want to use multi-op, I think we diverge the purpose though. I'm glad to start another pull request fall the other approach if you accept it as well. Or I can also make the changes required here and try to control the complexity but it will be always more complex than the other.

Reply:
in multi-op zookeeper approach, before creating znode can we check if it exists and use its result. Is that possible to do?

Reply:
Lets explore this approach further.. 
https://github.com/pravega/pravega/issues/3570#issuecomment-524562760
I think this will be easier to reason about as opposed to multi-op where the corner cases induced with concurrency are harder to tackle. 

So in recursive call we just want to create a new `collection` which is one higher than the previous one. 
And since we have a space of Integer.Max, even with multiple concurrent attempts to create a collection, we will not be in a situation where a new collection being created is immediately exhausted too. 
So your suggestion will work and is simple and clean. 
Do we see any problems with that?

Reply:
@shiveshr except the "immediately exhausted" problem theoretically,  I don't see any further problems with that.

I will give it another pass of consideration, and if no other concern I am gonna create another pull request follow that approach.

Reply:
i agree. lets open a new PR with just that change.

Reply:
Sorry for the late update. I give it another pass of consideration and cannot find further failure situation. Opened #4221